optional-option-result.md

```diff
@@ -0,0 +1,126 @@
+- Topic Name: Optional Option/Result
+- Start Date: 2022-06-25
+- RFC PR: [vlang/rfcs#00000](https://github.com/vlang/rfcs/pull/00000)
+- V Issue: [vlang/v#00000](https://github.com/vlang/v/issues/00000)
+
+# Summary
+
+The language currently does not provide a way to let developers handle memory allocation issues.
+Many developers are happy with the state of things: memory allocation failures are rare and a panic is acceptable.
+However the language could (and probably should) provide a way to handle these issues.
+
+A way to achieve this goal would be to return option/result without requiring the caller to handle failure scenarios.
+
+# Motivation
+
+V aims to provide developers the tools required to write solid applications however panicing on memory allocation can leave the application and its data in a invalid state.
+
+The developers should be able to handle these border cases without causing undue complexity on simpler applications.
+
+# Guide-level explanation
+
+Looking at this code:
+```v
+
+fn new_thingy () {
+ return &Thingy{
+  buffer: []int{len: 10000, cap: 30000}
+ }
+}
+
+```
+
+The code could fail as the underlying `malloc` code could fail as array is calling malloc which can panic
```

Comment:
Yes, this **is** a problem - but easily solved with `!` as stated above. For convenience reasons *V* does not enforce `!` nor `or{ ... }` at the call site at such scenarios (e.g. look at `mymap[]` which is also unsafe but you can make it safe by `mymap[...]or{5}` or `mymap[...]!`).

I would rather advocated introduction of a `-strict`/`-paranoid`/`-nopanic`/`-trulysafe`/`-pedantic` argument to the *V* compiler simply turning all `panic( ... )` calls into `return error( ... )` and thus forcing one to write `!` (or `or{}`) everywhere **explicitly**.

I.e. no new syntax, no new semantics. Only one argument and a very simple AST transformation.

Reply:
changing return type by passing extra cli argument is quite hard accomplish.
return type is always hardcoded in function definition. `return error(...)` requires `!` in return type.

Reply:
@ken0x0a I did not try to do any effort estimation but I am 100% certain it is doable and that it would not slow down the compiler much (if at all). *V*'s AST shall allow for such transformations already in its current form.